# Problem Set 1

### Question 1
What data type is each of the following?
 * 5
 * 5.0
 * 5 > 1
 * '5'
 * 5 * 2
 * '5' * 2
 * '5' + '2'
 * 5 / 2
 * 5 // 2
 * [5, 2, 1]
 * 5 in [1, 4, 6]
 * math.pi

In [1]:
#usage of type() function to get the type of each variable
import math
var = 5
print(var, "is of the type", type(var))
var = 5.0
print(var, "is of the type", type(var))
var = 5 > 1
print(var, "is of the type", type(var))
var = '5'
print(var, "is of the type", type(var))
var = 5 * 2
print(var, "is of the type", type(var))
var = '5' * 2
print(var, "is of the type", type(var))
var = '5' + '2'
print(var, "is of the type", type(var))
var = 5 / 2
print(var, "is of the type", type(var))
var = 5 // 2
print(var, "is of the type", type(var))
var = [5, 2, 1]
print(var, "is of the type", type(var))
var = 5 in [1, 4, 6]
print(var, "is of the type", type(var))
var = math.pi
print(var, "is of the type", type(var))

5 is of the type <class 'int'>
5.0 is of the type <class 'float'>
True is of the type <class 'bool'>
5 is of the type <class 'str'>
10 is of the type <class 'int'>
55 is of the type <class 'str'>
52 is of the type <class 'str'>
2.5 is of the type <class 'float'>
2 is of the type <class 'int'>
[5, 2, 1] is of the type <class 'list'>
False is of the type <class 'bool'>
3.141592653589793 is of the type <class 'float'>


### Question 2
Write (and evaluate) Python expressions that answer these questions:

##### a. 
How many letters are there in 'Supercalifragilisticexpialidocious'?

In [2]:
var = 'Supercalifragilisticexpialidocious'
#len() function to get the number of characters in a string
print('Number of letters in the word \"{}\" are {}.'.format(var,len(var))) 

Number of letters in the word "Supercalifragilisticexpialidocious" are 34.


##### b. 
Does 'Supercalifragilisticexpialidocious' contain 'ice' as a substring? 

In [3]:
var = 'Supercalifragilisticexpialidocious'
a = 'ice'
if a in var : #checks if the a is the substring of var
    print(var, "contains", a)

Supercalifragilisticexpialidocious contains ice


##### c. 
Which of the following words is the longest:
Supercalifragilisticexpialidocious, Honorificabilitudinitatibus, or
Bababadalgharaghtakamminarronnkonn? 

In [10]:
a = 'Supercalifragilisticexpialidocious'
b = 'Honorificabilitudinitatibus'
c = 'Bababadalgharaghtakamminarronnkonn'
lenA = len(a)
lenB = len(b)
lenC = len(c)
if lenA > lenB and lenA > lenC :
    print(a, 'is longest and the length is', lenA)
elif lenB > lenA and lenB > lenC :
    print(b, 'is longest and the length is', lenB)
else :
    print(c, 'is longest and the length is', lenC)

Bababadalgharaghtakamminarronnkonn is longest and the length is 34


##### d. 
Which composer comes first in the dictionary: 'Berlioz', 'Borodin', 'Brian',
'Bartok', 'Bellini', 'Buxtehude', 'Bernstein'. Which one comes last?

In [11]:
composerList = ['Berlioz', 'Borodin', 'Brian', 'Bartok', 'Bellini', 'Buxtehude', 'Bernstein']
composerList.sort()
print('sorted list is', composerList)
print('first word is', composerList[0])
print('The last word is', composerList[-1])

sorted list is ['Bartok', 'Bellini', 'Berlioz', 'Bernstein', 'Borodin', 'Brian', 'Buxtehude']
first word is Bartok
The last word is Buxtehude


### Question 3
##### a. 
Write a function inside(x,y,x1,y1,x2,y2) that returns True or False
depending on whether the point (x,y) lies in the rectangle with lower left
corner (x1,y1) and upper right corner (x2,y2).
* inside(1,1,0,0,2,3)
    >True
* inside(-1,-1,0,0,2,3)
    >False

In [16]:
def inside(x,y,x1,y1,x2,y2) : 
    'checks if the points (x,y) lies within the rectangle (x1,y1) and (x2,y2)'
    if (x >= x1 and x <= x2) and (y >= y1 and y <= y2) : #checks if the point lies within the rectangle
        print('The point ({},{}) lies within the rectangle ({},{}) and ({},{}).'.format(x,y,x1,y1,x2,y2))
        return True
    else:
        print('The point ({},{}) lies outside the rectangle ({},{}) and ({},{}).'.format(x,y,x1,y1,x2,y2))
        return False;
print(inside(1,1,0,0,2,3))
print(inside(-1,-1,0,0,2,3))

The point (1,1) lies within the rectangle (0,0) and (2,3).
True
The point (-1,-1) lies outside the rectangle (0,0) and (2,3).
False


#### b. 
Use function inside() from part a. to write an expression that tests whether
the point (1,1) lies in both of the following rectangles: one with lower left
corner (0.3, 0.5) and upper right corner (1.1, 0.7) and the other with lower
left corner (0.5, 0.2) and upper right corner (1.1, 2). 

In [13]:
def inside(x,y,x1,y1,x2,y2) : 
    'checks if the points (x,y) lies within the rectangle (x1,y1) and (x2,y2)'
    if (x >= x1 and x <= x2) and (y >= y1 and y <= y2) :
        return True
    else:
        return False;
a = inside(1,1,0.3,0.5,1.1,0.7)
b = inside(1,1,0.5,0.2,1.1,2)
if (a == True and b == True):
    print('The point (1,1) lies within both the rectangle (0.3,0.5)(1.1,0.7) and (0.5,0.2)(1.1,2).')
else:
    print('The point (1,1) does not lie within both the rectangle (0.3,0.5)(1.1,0.7) and (0.5,0.2)(1.1,2).')

The point (1,1) does not lie within both the rectangle (0.3,0.5)(1.1,0.7) and (0.5,0.2)(1.1,2).


### Question 4
You can turn a word into pig-Latin using the following two rules (simplified):
* If the word starts with a consonant, move that letter to the end and append 'ay'. For example, 'happy' becomes 'appyhay' and 'pencil' becomes 'encilpay'.
* If the word starts with a vowel, simply append 'way' to the end of the word. For example, 'enter' becomes 'enterway' and 'other' becomes 'otherway' . For our purposes, there are 5 vowels: a, e, i, o, u (so we count y as a consonant).

* Write a function pig() that takes a word (i.e., a string) as input and returns its pigLatin form. Your function should still work if the input word contains upper case
characters. Your output should always be lower case however.
* pig('happy')
>'appyhay'
* pig('Enter')
>'enterway'

In [18]:
def pig(s):
    stringLower = s.lower()
    a = stringLower[0]
    if a == 'i' or a == 'e' or a == 'i' or a == 'o' or a == 'u' : #If the first character is a vowel
        newString1 = stringLower + 'way'
        print(newString1)
    else: #If the first character is a consonant
        newString2 = stringLower + stringLower[0] + 'ay'
        print(newString2[1:])            
pig('happy')
pig('Enter')   
pig('Pencil')
pig('Other')

appyhay
enterway
encilpay
otherway


### Question 5
File bloodtype1.txt records blood-types of patients (A, B, AB, O or OO) at a clinic.
Write a function bldcount() that reads the file with name name and reports (i.e.,
prints) how many patients there are in each bloodtype.
* bldcount('bloodtype.txt')
> * There are 10 patients of blood type A.
> * There is one patient of blood type B.
> * There are 10 patients of blood type AB.
> * There are 12 patients of blood type O.
> * There are no patients of blood type OO.

In [13]:
def bldcount():
    'reads the file bloodtype1.txt and records the number of patients with the blood types'
    fileptr = open('bloodtype1.txt', 'r')
    typeList = fileptr.read()
    fileptr.close()
    bloodType = typeList.split()
    var = len(bloodType)
    bloodTypeList = ['A', 'B', 'AB', 'O', 'OO']
    count = [0 ,0 ,0 ,0 ,0] #blood group list in format [A, B, AB, O, OO]
    #For loop BEGIN; This compares the blood groups and keeps the count
    for i in range(var):
        if bloodType[i] == bloodTypeList[0]:
            count[0] += 1
        elif bloodType[i] == bloodTypeList[1]:
            count[1] += 1
        elif bloodType[i] == bloodTypeList[2]:
            count[2] += 1
        elif bloodType[i] == bloodTypeList[3]:
            count[3] += 1
        elif bloodType[i] == bloodTypeList[4]:
            count[4] += 1

    for i in range(len(bloodTypeList)):
        print('There are {} patients of blood type {}.'.format(count[i],bloodTypeList[i]))
          
bldcount()

There are 15 patients of blood type A.
There are 1 patients of blood type B.
There are 13 patients of blood type AB.
There are 15 patients of blood type O.
There are 0 patients of blood type OO.


### Question 6
Write a function curconv() that takes as input:
1. a currency represented using a string (e.g., 'JPY' for the Japanese Yen or
'EUR' for the Euro)
2. an amount

and then converts and returns the amount in US dollars.
* curconv('EUR', 100)
>122.96544
* curconv('JPY', 100)
>1.241401

In [2]:
def displayCurrencies():
    'Displays the currencies available in the file'
    fileptr = open('currencies.txt', 'r') #Open file in Read mode
    print('Currencies:')
    print('Acronym\tCurrent Value\tCurrency Name')
    print(fileptr.read())
    fileptr.close()

def curconv(a,b):
    'converts the currency'
    fileptr = open('currencies.txt', 'r') #Open file in Read mode
    listLines = fileptr.readlines() #Read lines
    fileptr.close()
    for i in listLines:
        if a in i:
            currencyLine = i
    chosenCur = currencyLine.split()
    value = chosenCur[1]
    convertedAmount = float(value) * float(b)
    return convertedAmount

#Display the available currencies
displayCurrencies()
#Ask to chose the required currency
currency = str(input('Enter the currency acronym (ex: AUD): '))
print('The currency you chose: {} '.format(currency))
#Ask for the amount to convert
amount = input('Enter the amount: ')
#Call function to convert currency
convertedAmount = curconv(currency,amount)
print('Converted Amount is: {}'.format(convertedAmount))

Currencies:
Acronym	Current Value	Currency Name
AUD	1.0345157	Australian Dollar
CHF	1.0237414	Swiss Franc
CNY	0.1550176	Chinese Yuan
DKK	0.1651442	Danish Krone
EUR	1.2296544	Euro
GBP	1.5550989	British Pound
HKD	0.1270207	Hong Kong Dollar
INR	0.0177643	Indian Rupee
JPY	0.01241401	Japanese Yen
MXN	0.0751848	Mexican Peso
MYR	0.3145411	Malaysian Ringgit
NOK	0.1677063	Norwegian Krone
NZD	0.8003591	New Zealand Dollar
PHP	0.0233234	Philippine Peso
SEK	0.148269	Swedish Krona
SGD	0.788871	Singapore Dollar
THB	0.0313789	Thai Baht

Enter the currency acronym (ex: AUD): AUD
The currency you chose: AUD 
Enter the amount: 100
Converted Amount is: 103.45157


### Question 7
Each of the following will cause an exception (an error). Identify what type of
exception each will cause.

* Q: Trying to add incompatible variables, as in adding 6 + ‘a’
* A: TypeError. It is because the addition is done on a string and an integer. 

* Q: Referring to the 12th item of a list that has only 10 items
* A: IndexError. This is because the list has only 10 items and the reference to the 12th item does not exist.

* Q: Using a value that is out of range for a function’s input, such as calling math.sqrt(-1.0)
* A: ValueError. The square root of a negative number is not defined in a simple form.

* Q: Using an undeclared variable, such as print(x) when x has not been defined
* A: NameError. It is because x is an unassigned identifier.

* Q: Trying to open a file that does not exist, such as mistyping the file name or looking in the wrong directory.
* A: FileNotFoundError. The error occurs as the file referred does not exist

### Question 8
Encryption is the process of hiding the meaning of a text by substituting letters in the
message with other letters, according to some system. If the process is successful, no
one but the intended recipient can understand the encrypted message. Cryptanalysis
refers to attempts to undo the encryption, even if some details of the encryption are
unknown (for example, if an encrypted message has been intercepted). The first step
of cryptanalysis is often to build up a table of letter frequencies in the encrypted text.
Assume that the string letters is already defined as
'abcdefghijklmnopqrstuvwxyz'. Write a function called frequencies()
that takes a string as its only parameter, and returns a list of integers, showing the
number of times each character appears in the text. Your function may ignore any
characters that are not in letters.
* frequencies('The quick red fox got bored and went home.')
>[1, 1, 1, 3, 5, 1, 1, 2, 1, 0, 1, 0, 1, 2, 4, 0, 1, 2, 0, 2,
1, 0, 1, 1, 0, 0]
* frequencies('apple')

In [17]:
def frequencies(s):
    'takes an input string and count the frequencties of each character in it'
    stringLower = s.lower()
    letters = 'abcdefghijklmnopqrstuvwxyz'
    lettersList = list(letters)
    count = [0] * 26
    for i in stringLower:
        for j in range(len(lettersList)):
            if i == lettersList[j]:
                count[j] += 1;
    print(count)
    
frequencies('The quick red fox got bored and went home.')
frequencies('apple')

[1, 1, 1, 3, 5, 1, 1, 2, 1, 0, 1, 0, 1, 2, 4, 0, 1, 2, 0, 3, 1, 0, 1, 1, 0, 0]
[1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]


### Question 9
The Sieve of Erastophenes is an algorithm -- known to ancient Greeks -- that finds
all prime numbers up to a given number n. It does this by first creating a list L from
2 to n and an (initially empty) list primeL. The algorithm then takes the first number
in list L (2) and appends it to list primeL, and then removes 2 and all its multiples
(4,6,8,10,12, ...) from L. The algorithm then takes the new first number in L (3) and
appends it to list primeL, and then removes from L 3 and all its remaining multiples
(9,15,21,...). So, in every iteration, the first number of list L is appended to list
primeL and then it and its multiples are removed from list L. The iterations stop
when list L becomes empty. Write a function sieve() that takes as input a positive
integer n, implements the above algorithm, and returns a list of all prime numbers up
to n.
* sieve(56)
>[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41,
43, 47, 53]
* sieve(368)
>[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41,
43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97,
101, 103, 107, 109, 113, 127, 131, 137, 139, 149,
151, 157, 163, 167, 173, 179, 181, 191, 193, 197,
199, 211, 223, 227, 229, 233, 239, 241, 251, 257,
263, 269, 271, 277, 281, 283, 293, 307, 311, 313,
317, 331, 337, 347, 349, 353, 359, 367]
* sieve(32)
>[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]

In [20]:
def sieve(n):
    'executes Sieve of Erastophenes algorithm'
    #create a list from 2 to n
    a = 2
    L = [] #Create an empty list L
    while(a <= n): #Loop to increment a from 2 to n and append it to the list L
        L.append(a)
        a += 1
    #print(L)
    primeL = [] #Empty list primeL
    for i in range(2 ,n + 1): #loop iterating the numbers in list L
        if i not in primeL and i in L: #Checks if i is in L and not in primeList. proceeds if both conditions are met.
            primeL.append(i) #appends i to the list primeL
            L.remove(i)    #removes i from L
            for j in range(i*i, n+1, i): #iterates for the multiples of i
                if (j%i == 0): #checks if j is divisible by i
                    if j in L: # only removes j from list L if it exists. 
                        L.remove(j)
   
    print('>>> sieve({})'.format(n))
    print(primeL)
  
sieve(56)
sieve(368)
sieve(32)

>>> sieve(56)
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53]
>>> sieve(368)
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367]
>>> sieve(32)
[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31]


### Question 10
Implement function triangleArea(a,b,c) that takes as input the lengths of the 3
sides of a triangle and returns the area of the triangle. By Heron's formula, the area
of a triangle with side lengths a, b, and c is
s(s - a)(s -b)(s -c)
, where
s = (a+b+c)/2.
* triangleArea(2,2,2)
>1.7320508075688772

In [22]:
import math
def triangleArea(a,b,c):
    'function to calculate the area of the triangle'
    s = (a+b+c)/2
    area = math.sqrt(s * (s-a) *(s-b) * (s-c))
    return area

area = triangleArea(2,2,2)
print('The area of the triangle is: {}'.format(area))

The area of the triangle is: 1.7320508075688772
